adal/src/main/java/com/microsoft/aad/adal/AuthenticationParameters.java

```diff
@@ -195,80 +192,354 @@ public void run() {
+            throws ResourceAuthenticationChallengeException {
         if (StringExtensions.isNullOrBlank(authenticateHeader)) {
             throw new ResourceAuthenticationChallengeException(AUTH_HEADER_MISSING);
-        } else {
-            Pattern p = Pattern.compile(REGEX);
-            Matcher m = p.matcher(authenticateHeader);
-
-            // If the header is in the right format, REGEX_VALUES will extract
-            // individual
-            // name-value pairs. This regex is not as exclusive, so it relies on
-            // the previous check to guarantee correctness:
-            if (m.matches()) {
-
-                // Get matching value pairs inside the header value
-                Pattern valuePattern = Pattern.compile(REGEX_VALUES);
-                String headerSubFields = authenticateHeader.substring(BEARER.length());
-                Logger.v(TAG + methodName, "Parse the header response. ", "Values in here:" + headerSubFields, null);
-                Matcher values = valuePattern.matcher(headerSubFields);
-                final Map<String, String> headerItems = new HashMap<>();
-                while (values.find()) {
-
-                    // values.group(0) is matching string
-                    if (!StringExtensions.isNullOrBlank(values.group(1))
-                            && !StringExtensions.isNullOrBlank(values.group(2))) {
-                        String key = values.group(1);
-                        String value = values.group(2);
-
-                        try {
-                            key = StringExtensions.urlFormDecode(key);
-                            value = StringExtensions.urlFormDecode(value);
-                        } catch (UnsupportedEncodingException e) {
-                            Logger.v(TAG + methodName, ADALError.ENCODING_IS_NOT_SUPPORTED.getDescription(), e.getMessage(), null);
-                        }
-
-                        key = key.trim();
-                        value = StringExtensions.removeQuoteInHeaderValue(value.trim());
-
-                        if (headerItems.containsKey(key)) {
-                            Logger.w(TAG + methodName,
-                                    "Key/value pair list contains redundant key. ",
-                                    "Redundant key: " + key,
-                                    ADALError.DEVELOPER_BEARER_HEADER_MULTIPLE_ITEMS);
-                        }
-
-                        headerItems.put(key, value);
-                    } else {
-                        // invalid format
-                        throw new ResourceAuthenticationChallengeException(AUTH_HEADER_INVALID_FORMAT);
-                    }
+        }
+
+        final List<Challenge> challenges = Challenge.parseChallenges(authenticateHeader);
+
+        // Grab the Bearer challenge
+        Challenge bearerChallenge = null;
+
+        for (final Challenge challenge : challenges) {
+            if (challenge.getScheme().equalsIgnoreCase(BEARER)) {
```

Comment:
you are assuming getScheme can never return null. I would flip the check around

Reply:
Resolved in 78b52908d59b9e8fbac42ce896314e58063224da